docs/proposals/error-handling-guideline.md

```diff
@@ -0,0 +1,279 @@
+```
+
+Suggested error message:
+
+```
+$ oras push --annotation-file annotation.json ghcr.io/library/alpine:3.9
+Error: invalid annotation json file: failed to load annotations from annotation.json.
+Annotation file doesn't match the required format. Please refer to the document at https://oras.land/docs/how_to_guides/manifest_annotations
+```
+
+#### Example 8: When pushing files, if the annotation value doesn't match the required syntax
+
+Current behavior and output:
+
+```
+$ oras push --annotation "key:value" ghcr.io/library/alpine:3.9
+Error: missing key in `--annotation` flag: key:value
+```
+
+Suggested error message:
+
+```
+$ oras push --annotation "key:value" ghcr.io/library/alpine:3.9
+Error: annotation value doesn't match the required format.
+Try `oras push --annotation "key=value" ghcr.io/library/alpine:3.9`  
+```
+
+#### Example 9: When failed to pull files from a public registry
+
+```
+$ oras pull docker.io/nginx:latest
+Error: failed to resolve latest: GET "https://registry-1.docker.io/v2/nginx/manifests/latest": response status code 401: unauthorized: authentication required: [map[Action:pull Class: Name:nginx Type:repository]]
+```
+
+Suggested error message:
+
+```
+$ oras pull docker.io/nginx:latest
+Error response from registry: pull access denied for docker.io/nginx:latest : response status code 401: unauthorized: requested access to the resource is denied
+Namespace is missing, do you mean `oras pull docker.io/library/nginx:latest`? 
+```
+
+#### Example 10: Neither registry nor OCI image layout provided when pushing a folder 
+
+Current behavior and output:
+
+```
+$ oras push /oras --format json
+Error: Head "https:///v2/oras/manifests/sha256:ffa50b27cd0096150c0338779c5090db41ba50d01179d851d68afa50b393c3a3": http: no Host in request URL
+```
+
+Suggested error message:
+
+```
+$ oras push /oras --format json
+Error: neither registry nor OCI image layout provided in the command 
+Usage: oras push [flags] <name>[:<tag>[,<tag>][...]] <file>[:<type>] [...]
+please specify a registry or an OCI image layout folder to push. Run "oras push -h" for more options and examples 
+```
+
+#### Example 11: Push a file or folder that doesn't exist
+
+Current behavior and output:
+
+```
+$ oras push localhost:5000/oras:v1 hello.txt
+Error: failed to stat /home/user/hello.txt: stat /home/user/hello.txt: no such file or directory
+```
+
+Suggested error message:
+
+```
+$ oras push localhost:5000/oras:v1 hello.txt
+Error: /home/user/hello.txt: no such file or directory
+```
+
+#### Example 12: Attach a file to an incompatible registry
+
+Current behavior and output:
+
+```
+$ oras attach docker.io/user/demo:v1 --artifact-type example/sbom sbom.spdx
+Uploading 6cb759c4296e sbom.spdx
+Uploaded  6cb759c4296e sbom.spdx
+Error: PUT "https://registry-1.docker.io/v2/user/demo/manifests/sha256:d2e409ed5674c9cf4fb5c13615288b130610ed893f97c78668ad233dc403976d": response status code 404: notfound: not found
```

Comment:
This is interesting that dockerhub returns 404 on PUT for not supporting referrers?

Reply:
Yes. Line260 is the original logs from Docker Hub, but it looks too simple and confusing.

Reply:
Dockerhub is returning 404 for not supporting manifests with subject field, **not** referrers API is not supported. [404 from listing referrers API call](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md#listing-referrers) is a known error. There is no doc or post saying Dockerhub will keep returning 404 until manifests with subject is supported, suggest remove this example.